src/Compilers/CSharp/Portable/Utilities/ValueSetFactory.NumericValueSet.cs

```diff
@@ -164,6 +164,17 @@ public IValueSet<T> Complement()
 
             public IValueSet<T> Intersect(IValueSet<T> o)
             {
+                // We use non-negative integers for Count/Length on types that list-patterns can be used on (ie.countable and indexable ones).
+                // But we need to upgrade them to regular integers to perform operations against full integer sets.
+                if (this is NumericValueSet<int, NonNegativeIntTC> nonNegativeThis && o is NumericValueSet<int, IntTC>)
```

Comment:
There's one place that a similar situation is handled (though this is a test while we want to track an evaluation):

https://github.com/dotnet/roslyn/blob/b0ebfe710c30df0303c8da0cdca64bcccdfe5bf8/src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs#L929-L933

I think a flag could find IsLengthOrCount=true and tweak the selected node with a matching IsLengthOrCount.